File: dotnet/StachExtensions/FactSet.Protobuf.Stach.Extensions/V3/StachV3JsonHelper.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;

namespace FactSet.Protobuf.Stach.Extensions.V3
{
    internal static class StachV3JsonHelper
    {
        public static TableSchema ParseTableSchema(JObject stachJObject)
        {
            var stachTable = new TableSchema();
            var stachV3TableJson = GetStachV3Properties(stachJObject, StachV3SchemaProperties.Table);

            if (!string.IsNullOrEmpty(stachV3TableJson))
            {
                var tableSchema = JObject.Parse(stachV3TableJson);
                if (tableSchema != null)
                {
                    var stachV3SplitResult = GetStachV3Properties(tableSchema, StachV3SchemaProperties.SplitResult);
                    if (!string.IsNullOrEmpty(stachV3SplitResult))
                    {
                        stachTable.SplitResult = JsonConvert.DeserializeObject<Dictionary<string, object>>(stachV3SplitResult);
                    }
                }
            }

            return stachTable;
        }

        public static TableSchema ParseViewsSchema(JObject stachJObject)
        {
            var customStachV3Views = new TableSchema();
            var stachV3ViewsJson = GetStachV3Properties(stachJObject, StachV3SchemaProperties.Views);

            if (!string.IsNullOrEmpty(stachV3ViewsJson))
            {
                var viewsArray = JArray.Parse(stachV3ViewsJson);
                if (viewsArray.Count > 0)
                {
                    var stachV3ViewObject = JObject.Parse(viewsArray[0].ToString());
                    customStachV3Views.ViewName = GetStachV3Properties(stachV3ViewObject, StachV3SchemaProperties.Name);

                    var viewColumns = GetStachV3Properties(stachV3ViewObject, StachV3SchemaProperties.Table, StachV3SchemaProperties.Columns);
                    if (!string.IsNullOrEmpty(viewColumns))
                    {
                        customStachV3Views.ViewColumns = JsonConvert.DeserializeObject<List<string>>(viewColumns);
                    }

                    var viewHeaderColumns = GetStachV3Properties(stachV3ViewObject, StachV3SchemaProperties.Table, StachV3SchemaProperties.Headers);
                    if (!string.IsNullOrEmpty(viewHeaderColumns))
                    {
                        customStachV3Views.ViewHeaders = JsonConvert.DeserializeObject<Dictionary<string, string>>(viewHeaderColumns);
                    }
                }
            }

            return customStachV3Views;
        }

        public static TableSchema ParseColumnsAndRows(JObject stachJObject)
        {
            var customStachV3Table = new TableSchema();

            var stachV3ColumnsJson = GetStachV3Properties(stachJObject, StachV3SchemaProperties.Columns);
            if (!string.IsNullOrEmpty(stachV3ColumnsJson))
            {
                customStachV3Table.Columns = JsonConvert.DeserializeObject<List<Dictionary<string, object>>>(stachV3ColumnsJson);
            }

            var stachV3RowsJson = GetStachV3Properties(stachJObject, StachV3SchemaProperties.Rows);
            if (!string.IsNullOrEmpty(stachV3RowsJson))
            {
                customStachV3Table.Rows = JsonConvert.DeserializeObject<List<Dictionary<string, string>>>(stachV3RowsJson);
            }

            return customStachV3Table;
        }

        public static TableSchema ParseMultiLevelHeaders(JObject stachJObject)
        {
            var multiLevelHeaders = new TableSchema();
            var stachV3MlhtJson = GetStachV3Properties(stachJObject, StachV3SchemaProperties.MultiLevelHeadersTable);

            if (!string.IsNullOrEmpty(stachV3MlhtJson))
            {
                var mlhtObject = JObject.Parse(stachV3MlhtJson);

                var mlhtColumns = GetStachV3Properties(mlhtObject, StachV3SchemaProperties.Columns);
                if (!string.IsNullOrEmpty(mlhtColumns))
                {
                    multiLevelHeaders.Columns = JsonConvert.DeserializeObject<List<Dictionary<string, object>>>(mlhtColumns);
                }

                var mlhtRows = GetStachV3Properties(mlhtObject, StachV3SchemaProperties.Rows);
                if (!string.IsNullOrEmpty(mlhtRows))
                {
                    multiLevelHeaders.Rows = JsonConvert.DeserializeObject<List<Dictionary<string, string>>>(mlhtRows);
                }
            }

            return multiLevelHeaders;
        }

        public static void AddGroupColumns(List<Dictionary<string, string>> rows, Dictionary<string, string> columnHeadersMapping)
        {
            var levels = GetLevels(rows);
            if (levels != null && levels.Any())
            {
                var maxLevel = levels.Max();
                for (int i = 0; i <= maxLevel; i++)
                {
                    var levelColumnName = $"{StachV3Constants.GroupPrefix}{i}";
                    if (!columnHeadersMapping.ContainsKey(levelColumnName))
                    {
                        columnHeadersMapping.Add(levelColumnName, levelColumnName);
                    }
                }
            }
        }

        public static Dictionary<string, string> BuildConcatenatedColumns(
            TableSchema multiLevelHeaders,
            TableSchema stachTable,
            TableSchema customStachV3Views)
        {
            var concatColumnsList = new Dictionary<string, string>();

            if (multiLevelHeaders?.Rows == null || multiLevelHeaders?.Columns == null)
            {
                return concatColumnsList;
            }

            for (int rowIndex = 0; rowIndex < multiLevelHeaders.Rows.Count; rowIndex++)
            {
                var concatColumnValue = string.Empty;
                var primaryKeyColumn = string.Empty;

                for (int colIndex = 0; colIndex < multiLevelHeaders.Columns.Count; colIndex++)
                {
                    var columnObject = multiLevelHeaders.Columns[colIndex][StachV3SchemaProperties.Name]?.ToString();
                    if (string.IsNullOrEmpty(columnObject))
                    {
                        continue;
                    }

                    var headerColumnValue = multiLevelHeaders.Rows[rowIndex].ContainsKey(columnObject)
                        ? multiLevelHeaders.Rows[rowIndex][columnObject]
                        : null;

                    var mlhtReferenceName = stachTable?.SplitResult?.ContainsKey(StachV3SchemaProperties.MultiLevelHeadersTableReference) == true
                        ? stachTable.SplitResult[StachV3SchemaProperties.MultiLevelHeadersTableReference]?.ToString()
                        : string.Empty;

                    if (columnObject == mlhtReferenceName)
                    {
                        primaryKeyColumn = headerColumnValue;
                    }

                    if (!string.IsNullOrEmpty(headerColumnValue))
                    {
                        if (customStachV3Views?.ViewHeaders?.ContainsKey(headerColumnValue) == true)
                        {
                            headerColumnValue = customStachV3Views.ViewHeaders[headerColumnValue] ?? headerColumnValue;
                        }
                        concatColumnValue = headerColumnValue + StachV3Constants.ColumnSeparator + concatColumnValue;
                    }
                }

                if (!string.IsNullOrEmpty(primaryKeyColumn))
                {
                    concatColumnsList[primaryKeyColumn] = concatColumnValue.TrimEnd(StachV3Constants.ColumnSeparator.ToCharArray());
                }
            }

            return concatColumnsList;
        }

        public static void MapColumnHeaders(
            TableSchema customStachV3Table,
            TableSchema customStachV3Views,
            Dictionary<string, string> concatColumnsList,
            Dictionary<string, string> columnHeadersMapping)
        {
            if (customStachV3Table?.Columns == null)
            {
                return;
            }

            for (int i = 0; i < customStachV3Table.Columns.Count; i++)
            {
                if (!customStachV3Table.Columns[i].ContainsKey(StachV3SchemaProperties.Name))
                {
                    continue;
                }

                var columnObject = customStachV3Table.Columns[i][StachV3SchemaProperties.Name]?.ToString();
                if (string.IsNullOrEmpty(columnObject))
                {
                    continue;
                }

                var headerName = columnObject;

                if (concatColumnsList?.ContainsKey(columnObject) == true)
                {
                    headerName = concatColumnsList[columnObject];
                }
                else if (customStachV3Views?.ViewHeaders?.ContainsKey(columnObject) == true)
                {
                    headerName = customStachV3Views.ViewHeaders[columnObject] ?? columnObject;
                }

                columnHeadersMapping[columnObject] = headerName;
            }
        }

        public static void CreateDataTableColumns(DataTable dataTable, Dictionary<string, string> columnHeadersMapping)
        {
            dataTable.Columns.AddRange(
                columnHeadersMapping.Select(kvp => new DataColumn(kvp.Value, typeof(string))).ToArray()
            );
        }

        public static void PopulateDataTableRows(
            DataTable dataTable,
            TableSchema customStachV3Table,
            Dictionary<string, string> columnHeadersMapping)
        {
            if (customStachV3Table?.Rows == null)
            {
                return;
            }

            for (int rowIndex = 0; rowIndex < customStachV3Table.Rows.Count; rowIndex++)
            {
                var dataRow = dataTable.NewRow();
                var rowDict = customStachV3Table.Rows[rowIndex];

                var securityNameColumnData = GetSecurityNameColumnData(rowDict, columnHeadersMapping);
                var levelIndex = GetLevelIndex(rowDict, columnHeadersMapping);

                foreach (var column in columnHeadersMapping)
                {
                    if (column.Key.StartsWith(StachV3Constants.GroupPrefix))
                    {
                        PopulateGroupColumn(dataTable, dataRow, column.Key, levelIndex, rowIndex, securityNameColumnData);
                    }
                    else
                    {
                        PopulateDataColumn(dataRow, rowDict, column);
                    }
                }

                dataTable.Rows.Add(dataRow);
            }
        }

        public static string GetSecurityNameColumnData(
            Dictionary<string, string> rowDict,
            Dictionary<string, string> columnHeadersMapping)
        {
            var securityColumn = columnHeadersMapping.FirstOrDefault(x =>
                x.Value.IndexOf(StachV3Constants.SecurityNameColumn, StringComparison.OrdinalIgnoreCase) >= 0);

            if (!string.IsNullOrEmpty(securityColumn.Key) && rowDict.ContainsKey(securityColumn.Key))
            {
                return rowDict[securityColumn.Key] ?? string.Empty;
            }

            return string.Empty;
        }

        public static int GetLevelIndex(
            Dictionary<string, string> rowDict,
            Dictionary<string, string> columnHeadersMapping)
        {
            var levelColumn = columnHeadersMapping.FirstOrDefault(x =>
                x.Key.IndexOf(StachV3Constants.LevelNameColumn, StringComparison.OrdinalIgnoreCase) >= 0);

            if (!string.IsNullOrEmpty(levelColumn.Key) && rowDict.ContainsKey(levelColumn.Key))
            {
                if (int.TryParse(rowDict[levelColumn.Key], out int levelIndex))
                {
                    return levelIndex;
                }
            }

            return 0;
        }

        public static void PopulateGroupColumn(
            DataTable dataTable,
            DataRow dataRow,
            string columnName,
            int levelIndex,
            int rowIndex,
            string securityNameColumnData)
        {
            if (int.TryParse(columnName.Replace(StachV3Constants.GroupPrefix, string.Empty), out int groupIndex))
            {
                if (levelIndex == groupIndex)
                {
                    dataRow[columnName] = securityNameColumnData;
                }
                else if (rowIndex > 0 && levelIndex > 0 && groupIndex < levelIndex)
                {
                    dataRow[columnName] = dataTable.Rows[rowIndex - 1][columnName];
                }
            }
        }

        public static void PopulateDataColumn(
            DataRow dataRow,
            Dictionary<string, string> rowDict,
            KeyValuePair<string, string> column)
        {
            var columnData = rowDict.ContainsKey(column.Key) ? rowDict[column.Key] : null;
            var columnName = column.Value;

            if (columnData != null && !string.IsNullOrEmpty(columnName))
            {
                dataRow[columnName] = columnData;
            }
        }

        public static void RemoveIgnoredColumns(DataTable dataTable)
        {
            foreach (var column in StachV3Constants.IgnoredColumns)
            {
                if (dataTable.Columns.Contains(column))
                {
                    dataTable.Columns.Remove(column);
                }
            }
        }

        public static long[] GetLevels(List<Dictionary<string, string>> rows)
        {
            if (rows == null || rows.Count == 0)
            {
                return Array.Empty<long>();
            }

            var levels = new HashSet<long>();

            foreach (var row in rows)
            {
                if (row.ContainsKey(StachV3Constants.LevelNameColumn) &&
                    long.TryParse(row[StachV3Constants.LevelNameColumn], out long level))
                {
                    levels.Add(level);
                }
            }

            return levels.ToArray();
        }

        public static string GetStachV3Properties(JObject source, string property)
        {
            if (source.TryGetValue(property, out JToken token) && token != null)
            {
                return token.ToString();
            }
            return string.Empty;
        }

        public static string GetStachV3Properties(JObject source, string property, string subProperty)
        {
            if (source.TryGetValue(property, out JToken token) && token != null)
            {
                var subToken = token[subProperty];
                if (subToken != null)
                {
                    return subToken.ToString();
                }
            }
            return string.Empty;
        }
    }
}

File: java/src/main/java/com/factset/protobuf/stach/extensions/v3/StachV3JsonHelper.java
package com.factset.protobuf.stach.extensions.v3;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

public class StachV3JsonHelper extends StachV3Constants {

    // JSON-specific fields
    public LinkedHashMap<String, String> columnHeadersMapping = new LinkedHashMap<>();
    public List<Map<String, String>> dataTable = new ArrayList<>();
    public LinkedHashMap<String, String> concatColumnsList = new LinkedHashMap<>();
    public List<List<Map<String, String>>> dataSet = new ArrayList<>();
    public TableSchema jsonStachTable = new TableSchema();
    public TableSchema customStachV3Table = new TableSchema();
    public TableSchema customStachV3Views = new TableSchema();
    public TableSchema jsonMultiLevelHeaders = new TableSchema();

    // Parses the table metadata (SplitResult) from the stach JSON object
    public void parseTableMetadata(JSONObject stachJObject) throws Exception {
        String stachV3TableJson = getStachV3Properties(stachJObject, StachV3SchemaProperties.Table);
        if (!stachV3TableJson.isEmpty()) {
            JSONObject tableSchema = new JSONObject(stachV3TableJson);
            String stachV3SplitResultJson = getStachV3Properties(tableSchema, StachV3SchemaProperties.SplitResult);
            if (!stachV3SplitResultJson.isEmpty()) {
                ObjectMapper mapper = new ObjectMapper();
                jsonStachTable.SplitResult = mapper.readValue(
                        tableSchema.get(StachV3SchemaProperties.SplitResult).toString(),
                        Map.class
                );
            }
        }
    }

    // Parses the views section (view name, columns, headers) from the stach JSON object
    public void parseViews(JSONObject stachJObject) throws Exception {
        String stachV3ViewsJson = getStachV3Properties(stachJObject, StachV3SchemaProperties.Views);
        if (stachV3ViewsJson.isEmpty()) {
            return;
        }

        JSONArray viewsArray = new JSONArray(stachV3ViewsJson);
        if (viewsArray.length() == 0) {
            return;
        }

        JSONObject stachV3ViewObject = new JSONObject(viewsArray.get(0).toString());
        ObjectMapper mapper = new ObjectMapper();

        customStachV3Views.ViewName = getStachV3Properties(stachV3ViewObject, StachV3SchemaProperties.Name);

        String viewColumns = getStachV3Properties(stachV3ViewObject, StachV3SchemaProperties.Table, StachV3SchemaProperties.Columns);
        if (!viewColumns.isEmpty()) {
            customStachV3Views.ViewColumns = mapper.readValue(viewColumns, List.class);
        }

        String viewHeaderColumns = getStachV3Properties(stachV3ViewObject, StachV3SchemaProperties.Table, StachV3SchemaProperties.Headers);
        if (!viewHeaderColumns.isEmpty()) {
            customStachV3Views.ViewHeaders = mapper.readValue(viewHeaderColumns, Map.class);
        }
    }

    // Parses columns and rows from the stach JSON object; also populates group-level columns
    public void parseColumnsAndRows(JSONObject stachJObject) throws Exception {
        ObjectMapper mapper = new ObjectMapper();

        String stachV3ColumnsJson = getStachV3Properties(stachJObject, StachV3SchemaProperties.Columns);
        if (!stachV3ColumnsJson.isEmpty()) {
            customStachV3Table.Columns = mapper.readValue(stachV3ColumnsJson, List.class);
        }

        String stachV3RowsJson = getStachV3Properties(stachJObject, StachV3SchemaProperties.Rows);
        if (!stachV3RowsJson.isEmpty()) {
            List<Map<String, Object>> rawRows = mapper.readValue(stachV3RowsJson, new TypeReference<List<Map<String, Object>>>() {});
            customStachV3Table.Rows = rawRows.stream()
                    .map(row -> {
                        Map<String, String> strRow = new LinkedHashMap<>();
                        row.forEach((k, v) -> strRow.put(k, v != null ? v.toString() : ""));
                        return strRow;
                    })
                    .collect(Collectors.toList());
        }

        long[] levels = getLevels(customStachV3Table.Rows);
        if (levels != null && levels.length > 0) {
            long maxLevel = Arrays.stream(levels).max().getAsLong();
            for (int i = 0; i <= maxLevel; i++) {
                String levelColumnName = groupPrefix + i;
                columnHeadersMapping.putIfAbsent(levelColumnName, levelColumnName);
            }
        }
    }

    // Parses the multi-level headers table (columns and rows) from the stach JSON object
    public void parseMultiLevelHeaders(JSONObject stachJObject) throws Exception {
        String stachV3MlhtJson = getStachV3Properties(stachJObject, StachV3SchemaProperties.MultiLevelHeadersTable);
        if (stachV3MlhtJson.isEmpty()) {
            return;
        }

        JSONObject mlhtObject = new JSONObject(stachV3MlhtJson);
        ObjectMapper mapper = new ObjectMapper();

        String mlhtColumns = getStachV3Properties(mlhtObject, StachV3SchemaProperties.Columns);
        if (!mlhtColumns.isEmpty()) {
            jsonMultiLevelHeaders.Columns = mapper.readValue(mlhtColumns, List.class);
        }

        String mlhtRows = getStachV3Properties(mlhtObject, StachV3SchemaProperties.Rows);
        if (!mlhtRows.isEmpty()) {
            jsonMultiLevelHeaders.Rows = mapper.readValue(mlhtRows, List.class);
        }
    }

    // Builds concatColumnsList by iterating multi-level header rows and concatenating header values
    public void buildConcatColumnsList() {
        if (jsonMultiLevelHeaders.Rows == null || jsonMultiLevelHeaders.Columns == null) {
            return;
        }

        for (int rowIndex = 0; rowIndex < jsonMultiLevelHeaders.Rows.size(); rowIndex++) {
            String concatColumnValue = "";
            String primaryKeyColumn = "";

            for (int colIndex = 0; colIndex < jsonMultiLevelHeaders.Columns.size(); colIndex++) {
                Map<String, Object> columnObj = jsonMultiLevelHeaders.Columns.get(colIndex);
                String columnObject = columnObj.get(StachV3SchemaProperties.Name).toString();
                if (columnObject == null) {
                    continue;
                }

                Map<String, String> rowObj = jsonMultiLevelHeaders.Rows.get(rowIndex);
                String headerColumnValue = rowObj.get(columnObject);

                String mlhtReferenceName = jsonStachTable.SplitResult != null
                        ? jsonStachTable.SplitResult.getOrDefault(StachV3SchemaProperties.MultiLevelHeadersTableReference, "").toString()
                        : "";
                if (columnObject.equals(mlhtReferenceName)) {
                    primaryKeyColumn = headerColumnValue;
                }

                if (headerColumnValue != null && !headerColumnValue.isEmpty()) {
                    if (customStachV3Views != null && customStachV3Views.ViewHeaders != null
                            && customStachV3Views.ViewHeaders.containsKey(headerColumnValue)) {
                        headerColumnValue = customStachV3Views.ViewHeaders.get(headerColumnValue);
                    }
                    concatColumnValue = headerColumnValue + columnSeparator + concatColumnValue;
                }
            }

            if (primaryKeyColumn != null && !primaryKeyColumn.isEmpty()) {
                concatColumnsList.put(primaryKeyColumn, concatColumnValue.replaceAll("\\s*\\|\\s*$", ""));
            }
        }
    }

    // Builds columnHeadersMapping by resolving each column name against view headers or concatColumnsList
    public void buildColumnHeadersMapping() {
        if (customStachV3Table.Columns == null) {
            return;
        }

        for (Map<String, Object> colObj : customStachV3Table.Columns) {
            if (!colObj.containsKey(StachV3SchemaProperties.Name)) {
                continue;
            }
            String columnObject = colObj.get(StachV3SchemaProperties.Name).toString();
            if (columnObject == null || columnObject.isEmpty()) {
                continue;
            }

            String headerName = columnObject;
            if (!concatColumnsList.isEmpty() && concatColumnsList.containsKey(columnObject)) {
                headerName = concatColumnsList.get(columnObject);
            } else if (customStachV3Views != null && customStachV3Views.ViewHeaders != null
                    && customStachV3Views.ViewHeaders.containsKey(columnObject)) {
                headerName = customStachV3Views.ViewHeaders.get(columnObject);
            }
            columnHeadersMapping.put(columnObject, headerName);
        }
    }

    // Iterates data rows and populates dataTable, including group hierarchy columns
    public void buildDataTable() {
        if (customStachV3Table.Rows == null) {
            return;
        }

        for (int rowIndex = 0; rowIndex < customStachV3Table.Rows.size(); rowIndex++) {
            Map<String, String> rowDict = customStachV3Table.Rows.get(rowIndex);
            Map<String, String> dataRow = new java.util.HashMap<>();

            String securityColumnKey = columnHeadersMapping.entrySet().stream()
                    .filter(x -> x.getValue().toLowerCase().contains(securityNameColumn.toLowerCase()))
                    .map(Map.Entry::getKey)
                    .findFirst().orElse("");
            String securityNameColumnData = (!securityColumnKey.isEmpty() && rowDict.containsKey(securityColumnKey))
                    ? rowDict.getOrDefault(securityColumnKey, "")
                    : "";

            String levelColumnKey = columnHeadersMapping.entrySet().stream()
                    .filter(x -> x.getKey().toLowerCase().contains(levelNameColumn.toLowerCase()))
                    .map(Map.Entry::getKey)
                    .findFirst().orElse("");
            int levelIndex = 0;
            if (!levelColumnKey.isEmpty() && rowDict.containsKey(levelColumnKey)) {
                try {
                    levelIndex = Integer.parseInt(rowDict.get(levelColumnKey));
                } catch (Exception e) {
                    levelIndex = 0;
                }
            }

            populateDataRow(dataRow, rowDict, rowIndex, levelIndex, securityNameColumnData);
            dataTable.add(dataRow);
        }
    }

    // Fills a single data row with regular column values and group hierarchy columns
    protected void populateDataRow(Map<String, String> dataRow, Map<String, String> rowDict,
                                   int rowIndex, int levelIndex, String securityNameColumnData) {
        for (Map.Entry<String, String> column : columnHeadersMapping.entrySet()) {
            if (!column.getKey().contains(groupPrefix)) {
                String columnData = rowDict.getOrDefault(column.getKey(), "");
                String columnName = column.getValue();
                if (columnName == null || columnName.isEmpty()) {
                    continue;
                }
                dataRow.put(columnName, columnData);
            } else {
                String columnName = column.getKey();
                int groupIndex = 0;
                try {
                    groupIndex = Integer.parseInt(columnName.replace(groupPrefix, ""));
                } catch (Exception e) {
                    groupIndex = 0;
                }

                if (levelIndex == groupIndex) {
                    dataRow.put(columnName, securityNameColumnData);
                } else if (rowIndex > 0 && levelIndex > 0 && groupIndex < levelIndex) {
                    dataRow.put(columnName, dataTable.get(rowIndex - 1).get(columnName));
                }
            }
        }
    }

    // Helper to get levels
    protected static long[] getLevels(List<Map<String, String>> rows) {
        List<Long> levels = new ArrayList<>();
        if (rows == null || rows.isEmpty()) {
            return new long[0];
        }
        for (Map<String, String> row : rows) {
            if (!row.containsKey("level")) {
                continue;
            }
            String columnX = row.get("level");
            if (columnX != null && !columnX.isEmpty()) {
                try {
                    long level = Long.parseLong(columnX);
                    if (!levels.contains(level)) {
                        levels.add(level);
                    }
                } catch (Exception e) {
                    // ignore parse error
                }
            }
        }
        return levels.stream().mapToLong(Long::longValue).toArray();
    }

    // Helper to get stach v3 json properties
    protected static String getStachV3Properties(JSONObject source, String property) throws JSONException {
        if (source.has(property) && source.get(property) != null) {
            return source.get(property).toString();
        }
        return "";
    }

    // Helper to get stach v3 json sub-properties
    protected static String getStachV3Properties(JSONObject source, String property, String subProperty) throws JSONException {
        if (source.has(property) && source.get(property) != null) {
            JSONObject token = source.getJSONObject(property);
            if (token.has(subProperty) && token.get(subProperty) != null) {
                return token.get(subProperty).toString();
            }
        }
        return "";
    }
}